core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v1.go

```diff
@@ -0,0 +1,391 @@
+func (q *upkeepLogQueue) enqueue(blockThreshold int64, logsToAdd ...logpoller.Log) (int, int) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	logs := q.logs
+	var added int
+	for _, log := range logsToAdd {
+		if log.BlockNumber < blockThreshold {
+			// q.lggr.Debugw("Skipping log from old block", "blockThreshold", blockThreshold, "logBlock", log.BlockNumber, "logIndex", log.LogIndex)
+			continue
+		}
+		logid := logID(log)
+		if _, ok := q.visited[logid]; ok {
+			// q.lggr.Debugw("Skipping known log", "blockThreshold", blockThreshold, "logBlock", log.BlockNumber, "logIndex", log.LogIndex)
+			continue
+		}
+		added++
+		logs = append(logs, log)
+		q.visited[logid] = log.BlockNumber
+	}
+	q.logs = logs
+
+	var dropped int
+	if added > 0 {
+		dropped = q.clean(blockThreshold)
+		q.lggr.Debugw("Enqueued logs", "added", added, "dropped", dropped, "blockThreshold", blockThreshold, "q size", len(q.logs), "visited size", len(q.visited))
+	}
+
+	prommetrics.AutomationLogBufferFlow.WithLabelValues(prommetrics.LogBufferFlowDirectionIngress).Add(float64(added))
+	prommetrics.AutomationLogBufferFlow.WithLabelValues(prommetrics.LogBufferFlowDirectionDropped).Add(float64(dropped))
+
+	return added, dropped
+}
+
+// clean removes logs that are older than blockThreshold and drops logs if the limit for the
+// given upkeep was exceeded. Returns the number of logs that were dropped.
+func (q *upkeepLogQueue) clean(blockThreshold int64) int {
+	blockRate := int(q.opts.blockRate.Load())
+	maxLogsPerWindow := int(q.opts.logLimitHigh.Load())
+
+	// sort logs by block number, tx hash and log index
+	// to keep the q sorted and to ensure that logs can be
+	// grouped by block windows for the cleanup
+	sort.SliceStable(q.logs, func(i, j int) bool {
+		return LogSorter(q.logs[i], q.logs[j])
+	})
+	// cleanup logs that are older than blockThreshold
+	// and drop logs if the window/s limit for the given upkeep was exceeded
+	updated := make([]logpoller.Log, 0)
+	var dropped, expired, currentWindowCapacity int
+	var currentWindowStart int64
+	for _, l := range q.logs {
+		if blockThreshold > l.BlockNumber { // old log, removed
+			prommetrics.AutomationLogBufferFlow.WithLabelValues(prommetrics.LogBufferFlowDirectionExpired).Inc()
+			// q.lggr.Debugw("Expiring old log", "blockNumber", l.BlockNumber, "blockThreshold", blockThreshold, "logIndex", l.LogIndex)
+			logid := logID(l)
+			delete(q.visited, logid)
+			expired++
+			continue
+		}
+		start, _ := getBlockWindow(l.BlockNumber, blockRate)
+		if start != currentWindowStart {
+			// new window, reset capacity
+			currentWindowStart = start
+			currentWindowCapacity = 0
+		}
+		currentWindowCapacity++
+		// if capacity has been reached, drop the log
+		if currentWindowCapacity > maxLogsPerWindow {
+			prommetrics.AutomationLogBufferFlow.WithLabelValues(prommetrics.LogBufferFlowDirectionDropped).Inc()
+			// TODO: check if we should clean visited as well, so it will be possible to add the log again
```

Comment:
is this needed? if so can you add a ticket to track it, if not remove the comment

Reply:
changed `visited` to `states`, now we update the state to `dropped`